website/blog/deployers.md

```diff
@@ -0,0 +1,610 @@
+    return &protos.GetListenerAddressReply{Address: "localhost:0"}, nil
+}
+
+func (h *handler) ExportListener(context.Context, *protos.ExportListenerRequest) (*protos.ExportListenerReply, error) {
+    return &protos.ExportListenerReply{}, nil
+}
+```
+
+<div class="note">
+<code>weaver multi</code>'s implementation of <code>GetListenerAddress</code> always returns <code>"localhost:0"</code>. Its <code>ExportListener</code> runs a local HTTP proxy on the address specified in the <code>LocalAddress</code> field of the <a href="https://pkg.go.dev/github.com/ServiceWeaver/weaver#ListenerOptions">ListenerOptions</a> passed to <a href="https://pkg.go.dev/github.com/ServiceWeaver/weaver#Instance">Listener</a>. This proxy balances traffic across the addresses of the listeners reported to <code>ExportListener</code>. This balances traffic across all replicas of a listener. <code>weaver gke</code> implements listeners by configuring load balancers in Google Cloud.
+</div>
+
+### Telemetry
+
+Finally, we implement the telemetry methods. All logs produced by a weavelet are
+received by the `HandleLogEntry` method. Our deployer uses a pretty printer from
+Service Weaver's `logging` library to print the logs to stdout. Similarly, all
+traces produced by a weavelet are received by the `HandleTraceSpans` function.
+For simplicity, our deployer ignore traces.
+
+```golang
+// Responsibility 3: Telemetry.
+func (h *handler) HandleLogEntry(_ context.Context, entry *protos.LogEntry) error {
+    pp := logging.NewPrettyPrinter(colors.Enabled())
+    fmt.Println(pp.Format(entry))
+    return nil
+}
+
+func (h *handler) HandleTraceSpans(context.Context, []trace.ReadOnlySpan) error {
+    return nil
+}
+```
+
+<div class="note">
+<code>weaver multi</code> writes logs and traces to files. <code>weaver gke</code> exports logs and traces to <a href="https://cloud.google.com/logging">Cloud Logging</a> and <a href="https://cloud.google.com/trace">Cloud Trace</a>.
+</div>
+
+### Main
+
+Finally, we implement `main`. We create a deployer, spawn the main component,
+and block.
+
+```go
+func main() {
+    flag.Parse()
+    d := &deployer{weavelets: map[string]*weavelet{}}
+    d.spawn("main")
+    select {} // block forever
+}
+```
+
+## Advanced Deployer Features
+
+The multiprocess deployer in the previous section was designed to be as simple
+as possible. Real-world deployers, on the other hand, require a number of more
+advanced features. Enumerating and explaining how to implement these features is
+beyond the scope of this blog post, but we'll summarize some advanced features
+here. You can also review the implementations of our [`weaver
+multi`][weaver_github] and [`weaver gke`][weaver_gke_github] deployers as
+reference.
+
+- **Failure Detection**. A deployer should detect when a weavelet has failed and
+  notify the other weavelets accordingly. You can use the `Envelope.GetHealth`
```

Comment:
Note that GetHealth won't work if the machine hosting the weavelet and the envelope is dead.

Maybe mention that multi-machine deployers will have to do their own distributed health-checking to detect failures?